### PR TITLE
Jm partial function recovery

### PR DESCRIPTION
**Pull request by jost125**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -357,6 +357,46 @@ TrySafe::fromCallable(fn() => $getTheUltimateAnswerOrThrow(true))
  
 _Disclaimer:_ As you probably noticed, we reflect possible exception in return type now, but on the other side, we've lost the information that wrapped success value is `float`.  This applies also to `Option`, `ArrayList` etc. Unfortunately there is no silver bullet solution, until PHP have the generics implemented (but hey, have look at [phpstan generics templates](https://medium.com/@ondrejmirtes/generics-in-php-using-phpdocs-14e7301953)). 
 
+#### TrySafe recovery
+
+We have already learned, that `TrySafe` can server to chaining dependant operations that can fail (via `flatMap`). How about if we want to have some fallback / recovery in the middle of that chain?
+
+This is where `recover*` methods come in handy. Let's take a look on this example:
```
Comment (honzatrtik) [inline README.md]: Take a look at

File: README.md [hunk 2]
```diff
@@ -357,6 +357,46 @@ TrySafe::fromCallable(fn() => $getTheUltimateAnswerOrThrow(true))
  
 _Disclaimer:_ As you probably noticed, we reflect possible exception in return type now, but on the other side, we've lost the information that wrapped success value is `float`.  This applies also to `Option`, `ArrayList` etc. Unfortunately there is no silver bullet solution, until PHP have the generics implemented (but hey, have look at [phpstan generics templates](https://medium.com/@ondrejmirtes/generics-in-php-using-phpdocs-14e7301953)). 
 
+#### TrySafe recovery
+
+We have already learned, that `TrySafe` can server to chaining dependant operations that can fail (via `flatMap`). How about if we want to have some fallback / recovery in the middle of that chain?
+
```
Comment (honzatrtik) [inline README.md]: Can serve for chaining? (But not sure about this one)

File: README.md [hunk 3]
```diff
@@ -357,6 +357,46 @@ TrySafe::fromCallable(fn() => $getTheUltimateAnswerOrThrow(true))
...
+This is where `recover*` methods come in handy. Let's take a look on this example:
+
+```php
+/** @var callable(Throwable): TrySage<Gps> */
+$recovery = fn (Throwable $ex): TrySafe => $backupApi->findGps($query);
```
Comment (honzatrtik) [inline README.md]: Typo `TrySage`

File: README.md [hunk 4]
```diff
@@ -357,6 +357,46 @@ TrySafe::fromCallable(fn() => $getTheUltimateAnswerOrThrow(true))
...
+
+/** @var TrySafe<int> */
+$distance = $api
+    ->findGps($query)
+    ->recoverWith($findGpsFromBackupApi)
+    ->map(fn (Gps $gps) => $this->getDistance($home));
+```
+
+There is three `recover*` methods:
```
Comment (honzatrtik) [inline README.md]: There are

File: README.md [hunk 5]
```diff
@@ -357,6 +357,46 @@ TrySafe::fromCallable(fn() => $getTheUltimateAnswerOrThrow(true))
...
+- `recover` - return directly value to which you want to recover
```
Comment (honzatrtik) [inline README.md]: Return value, you want computation to be recovered with

File: README.md [hunk 6]
```diff
@@ -357,6 +357,46 @@ TrySafe::fromCallable(fn() => $getTheUltimateAnswerOrThrow(true))
...
+- `recoverWith` - return value wrapped in `TrySafe`, this behaves similary like `flatMap`, but on failures. Therefore you can chain multiple things
+- `recoverWithPF` - checks where to apply the recovery by first closure or if it should keep the result intact. This can be handy if you want to recover only on specific exceptions or in specific circumstances.
```
Comment (honzatrtik) [inline README.md]: Recovers failure only if passed predicate evaluates to true

Or smthng like that?

File: src/Bonami/Collection/TrySafe.php [hunk 7]
```diff
@@ -216,6 +221,15 @@ public function recoverWith(callable $callable): TrySafe
                 })->flatMap($id);
             }
 
+            public function recoverWithPF(callable $case, callable $recovery): TrySafe
```
Comment (honzatrtik) [inline src/Bonami/Collection/TrySafe.php]: Somehow im not sure about introducing concept of pf into collections (do we mention it elsewhere?)
What about just `recoverWithIf` w/ the exact same semantics?

Comment (jost125) [reply src/Bonami/Collection/TrySafe.php]: Renamed

File: README.md (hunk 6 above)
Comment (jost125) [reply README.md]: :+1:

File: README.md (hunk 5 above)
Comment (jost125) [reply README.md]: End the sentence with **with** might be confusing with the line bellow, where I explain `recoverWith` operation semantics.

Comment (jost125) [reply README.md]: I have rewritten it entirely

File: README.md (hunk 4 above)
Comment (jost125) [reply README.md]: :+1:

File: README.md (hunk 3 above)
Comment (jost125) [reply README.md]: :+1:

File: README.md (hunk 2 above)
Comment (jost125) [reply README.md]: I have changed it to:

```
We have already learned, that `TrySafe` can be used for chaining dependant operations that can fail (via `flatMap`).
```

File: README.md (hunk 1 above)
Comment (jost125) [reply README.md]: :+1:

File: README.md (hunk 2 above)
Comment (honzatrtik) [reply README.md]: dependant -> dependent

**Review by honzatrtik — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (jost125) [reply README.md]: :+1:

**Pull request closed without merging**